PolyEngine/Core/Src/pe/core/utils/Result.hpp

```diff
@@ -0,0 +1,285 @@
+		constexpr T valueOr(U&& alt) const&
+		{
+			if (isOk())
+				return value();
+			else
+				return static_cast<T>(std::forward<U>(alt));
+
+		}
+
+		// Verstion of valueOr which is called for instances of Result where 
+		// stored valid value type (T) isn't copy constructible but instead 
+		// has to be move constructible.
+		// @tparam U - type of alternative value to return. U must be 
+		//		convertible to T
+		// @param alt - alternative value to return if this doesn't contain 
+		//		valid value
+		// @returns valid value if has valid value or given alternative value
+		//		if has error
+		template <typename U>
+		constexpr T valueOr(U&& alt) &&
+		{ 
+			if (isOk())
+				return std::move(value());
+			else
+				return static_cast<T>(std::forward<U>(alt));
+		}
+
+
+
+
+		// returns error if this stores an error, crashes otherwise :(
+		constexpr E& error() & { return std::get<E>(m_value); }
+		constexpr const E& error() const & { return std::get<E>(m_value); }
+		constexpr E&& error() && { return std::get<E>(std::move(m_value)); }
+
+	private:
+		std::variant<T, E> m_value;
+	};
+
+	//------------------------------------------------------------------------------
+	// Specialized result class for results where there is no result but
+	// there can be an error message.
+	template <typename E>
+	class Result<void, E> final : public BaseObjectLiteralType<>
+	{
+	public:
+		constexpr Result() = delete;
+		constexpr Result(Result&) = default;
+		constexpr Result(Result&&) = default;
+		constexpr Result& operator=(Result&) = default;
+		constexpr Result& operator=(Result&&) = default;
+
+		// Ctor taking ok value. In fact it doesn't take any value it's just 
+		// forces user to explicitly set state of this type of Result 
+		// instance.
+		// @param okVal - this is empty anyway
+		constexpr Result(Ok<void> okVal) {}
+
+		// Ctor taking error value. Object created with this ctor will return
+		// false from isOk and true from isErr.
+		// @param errVal - value for error/invalid result
+		constexpr Result(Err<E> errVal) : m_value(std::move(errVal).get()) {}
+
+		// Ctor taking error value. Object created with this ctor will return
+		// false from isOk and true from isErr. Using this ctor rather than
+		// the one taing Err<T> will save you one move operation.
+		// @param errVal - value for error/invalid result
+		explicit constexpr Result(E errVal) : m_value(std::move(errVal)) {}
+
+		// @returns true if result stores valid value, false if stores error
+		//		value
+		constexpr bool isOk() const { return m_value.has_value() == false; }
+
+		// @returns false if result stores valid value, true if stores error
+		//		value
+		constexpr bool isErr() const { return !isOk(); }
+
+
+
+		// returns error if this stores an error, crashes otherwise :(
+		constexpr E& error() & { return m_value.value(); }
+		constexpr const E& error() const & { return m_value.value(); }
+		constexpr E&& error() && { return std::move(m_value).value(); }
```

Comment:
Return by value.

Reply:
rvo works only for variables which are created within a method and then returned